sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/PubsubIOJsonTable.java

```diff
@@ -0,0 +1,193 @@
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.sdk.extensions.sql.BeamSqlTable;
+import org.apache.beam.sdk.extensions.sql.impl.schema.BeamIOType;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubIO;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.util.JsonToRowUtils;
+import org.apache.beam.sdk.util.RowJsonDeserializer;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.Row;
+
+/**
+ * <i>Experimental</i>
+ *
+ * <p>Wraps the {@link PubsubIO} with JSON messages into {@link BeamSqlTable}.
+ *
+ * <p>This enables {@link PubsubIO} registration in Beam SQL environment as a table, including DDL
+ * support.
+ *
+ * <p>For example:
+ * <pre>
+ *
+ *  CREATE TABLE topic (name VARCHAR, age INTEGER)
+ *     TYPE 'pubsub'
+ *     LOCATION projects/&lt;GCP project id&gt;/topics/&lt;topic name&gt;
+ *     TBLPROPERTIES '{ \"timestampAttributeKey\" : &lt;timestamp attribute&gt; }';
+ *
+ *   SELECT name, age FROM topic;
+ *
+ * </pre>
+ */
+@AutoValue
+@Internal
+@Experimental
+abstract class PubsubIOJsonTable implements BeamSqlTable, Serializable {
+
+  /**
+   * Schema of the pubsubs message payload.
+   *
+   * <p>Only UTF-8 flat JSON objects are supported at the moment.
+   */
+  abstract Schema getPayloadSchema();
+
+  /**
+   * Attribute key of the Pubsub message from which to extract the event timestamp.
+   *
+   * <p>This attribute has to conform to the same requirements as in {@link
+   * PubsubIO.Read.Builder#withTimestampAttribute}.
+   *
+   * <p>Short version: it has to be either millis since epoch or string in RFC 3339 format.
+   */
+  abstract String getTimestampAttribute();
+
+  /**
+   * Pubsub topic name.
+   *
+   * <p>Topic is the only way to specify the Pubsub source. Explicitly specifying the subscription
+   * is not supported at the moment. Subscriptions are automatically created an managed.
+   */
+  abstract String getTopic();
+
+  static Builder builder() {
+    return new AutoValue_PubsubIOJsonTable.Builder();
+  }
+
+  /**
+   * Table schema.
+   *
+   * <p>Inherited from {@link BeamSqlTable}. Different from {@link #getPayloadSchema()},
+   * includes timestamp attribute.
+   */
+   public abstract Schema getSchema();
+
+  @Override
+  public BeamIOType getSourceType() {
+    return BeamIOType.UNBOUNDED;
+  }
+
+  @Override
+  public PCollection<Row> buildIOReader(Pipeline pipeline) {
+    return
+        PBegin
+            .in(pipeline)
+            .apply("readFromPubsub", readMessagesWithAttributes())
+            .apply("parseJsonPayload", parseJsonPayload())
```

Comment:
<!--new_thread; commit:464b72bb1ce835937c456f661d5d90e86ae76c94; resolved:0-->
This looks like the only line you have to swap out if you switch from JSON to CSV to Avro to Proto, etc, yes?

Reply:
Yes. For protos and avros there are also shortcuts like `PubsubIO.readProtos()` and `PubsubIO.readAvros()` so it's probably going to be a little simpler